volumes.cpp

```diff
@@ -1301,18 +1321,156 @@ SurfaceVolume& makeSurface( const SurfaceCard* card, VolumeCache* v){
     else{
       throw std::runtime_error( mnemonic + " is not a supported surface" );
     }
-    
-    if( card->getTransform().hasData() ){
-      const Transform& transform = card->getTransform().getData();
-      surface->setTransform( &transform );
-    }
-    
-    cache.insert( card, surface );
-    return *surface;
-    
+  } 
+  if( card->getTransform().hasData() ){
+    const Transform& transform = card->getTransform().getData();
+    surface->setTransform( &transform );
   }
   
+  cache.insert( card, surface );
+  return *surface;
+    
+}
   
+
+SurfaceVolume* FacetSurface( const std::string mnemonic, const std::vector< double > args, int facet ){
+  if( mnemonic == "rcc" ){
+    return rccFacet( args, facet );
+  }
+  else if(mnemonic == "box"){
+    return boxFacet( args, facet );
+  }
+  else if( mnemonic == "rpp"){
+    return rppFacet( args, facet );
+  }
+  else if( mnemonic == "hex" || mnemonic == "rhp" ){
+    return hexFacet( args, facet );
+  }
+  else if( mnemonic == "rec" ){
+    return recFacet( args, facet );
+  }
+  throw std::runtime_error( mnemonic + " does not have macrobody facet support at this time." );
+}    
+
+
+//This function is used to find D in the planar equation Ax + By + Cz - D = 0 
+//when given two vectors.
+//The positioning vector always uses the first three arguments; i determines 
+//the first argument of the vector to which the plane will be parallel.  If 
+//end is true, the plane is at the end of the second vector in relation to the 
+//end of the first vector, otherwise it intersects the end of the first vector.
+double planePoint( std::vector< double > args, int i, bool end ){
+  if( end ){
+    return args.at(i)*( args.at(0) + args.at(i) ) + args.at(i + 1)*( args.at(1) + args.at(i + 1) ) + args.at(i + 2)*( args.at(2) + args.at(i + 2) );
+  }
+  else{
+    return args.at(i)*args.at(0) + args.at(i + 1)*args.at(1) + args.at(i + 2)*args.at(2);
+  }
+}
+
+
+SurfaceVolume* rccFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
+    //cylinder surface
+    return new RccVolume( Vector3d(args), Vector3d(args,3), args.at(6), true );
+  }
+  else if( facet == 2 || facet == 3 ){
+    //the two end planes
+    bool end = ( facet == 2);
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, end );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else{
+    throw std::runtime_error( "rcc only has 3 facets" );
+  }
+}
+
+SurfaceVolume* boxFacet( const std::vector< double > args, int facet ){
+  //This function creates the planes for the facets of box.
+  if( facet >= 1 && facet <= 6 ){
+    //if the facet number is odd, the plane is at the end of the vector specified
+    //which consists of arguments 3,4,5 for 1 & 2, 6,7,8 for 3 & 4, and so on.
+    int  i = ( ( facet + facet%2 ) / 2 ) * 3;
```

Comment:
I don't think that quite works, because it rounds down instead of up.  Using your math, facet 2 and 3 would have the same direction, when it needs to be 1 and 2, and the facet 6 would try to start on argument 12, when the last one is argument 11.
I can make a direction variable and use first_arg, though.